File: Problem64_FloodFill.java
// Time Complexity : O(m * n)
// Space Complexity : O(m * n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
    int val;
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        this.val = image[sr][sc];
        dfs(image, sr, sc, newColor);
        return image;
    }
    public void dfs(int[][] image, int row, int col, int newColor) {
        if(row < 0 || row >= image.length || col < 0 || col >= image[0].length || image[row][col] != val || image[row][col] == newColor)
            return;
        if(row >= 0 && row <= image.length - 1 && col >= 0 && col <= image[0].length - 1 && image[row][col] == val) {
            image[row][col] = newColor;
            dfs(image, row + 1, col, newColor);
            dfs(image, row - 1, col, newColor);
            dfs(image, row, col + 1, newColor);
            dfs(image, row, col - 1, newColor);
        }
    }
}

File: Problem-64 Flood Fill.py
# 733. Flood Fill
# https://leetcode.com/problems/flood-fill/

# Logic: Use DFS from the source given. Push the source in the stack and when you pop the same, 
# check its neighbours if they are of the same color push them to the stack and update it in visited. 
# Also update that cell with the new color.

# Time Complexiety: O(n)
# Space Complexiety: O(n)

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        n = len(image)
        m = len(image[0])
        color = image[sr][sc]
        stack = [(sr,sc)]
        neighbours = [(0,1),(0,-1),(1,0),(-1,0)]
        image[sr][sc] = newColor
        visited = {(sr,sc)}
        
        while stack:
            row, col = stack.pop()
            
            for neighbour in neighbours:
                newRow = row + neighbour[0]
                newCol = col + neighbour[1]
                
                if 0 <= newRow < n and 0 <= newCol < m:
                    if image[newRow][newCol] == color and (newRow,newCol) not in visited:
                        stack.append((newRow,newCol))
                        visited.add((newRow,newCol))
                        image[newRow][newCol] = newColor
                        
        return image